File: allegro_hand_teleop/scripts/allegro_hand_initial.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import JointState
from std_msgs.msg import Float64MultiArray

# ---------------- CLASS ----------------
class AllegroHandInitialNode():
    def __init__(self):
        print('AllegroHandInitialNode: initializing node')

        # Initialize Variables
        self.ready = False
        self.joint_names = [
            'jif1', 'jif2', 'jif3', 'jif4',
            'jmf1', 'jmf2', 'jmf3', 'jmf4',
            'jpf1', 'jpf2', 'jpf3', 'jpf4',
            'jth1', 'jth2', 'jth3', 'jth4'
        ]
        self.joint_states = None

        # ROS Infrastructure
        topic = 'joint_states'
        self.sub_joint_states = rospy.Subscriber(topic, JointState, self.joint_states_callback)
        topic = "allegro_hand_controller/command"
        self.command_pub = rospy.Publisher(topic, Float64MultiArray, queue_size=1)

        # Wait to get ready
        print('Waiting for joint_states...')
        while (not self.ready):
            pass
        print('Received first joint_states message')

        msg = Float64MultiArray()
        msg.data = self.joint_states.position

        rate = rospy.Rate(1000) # 10 Hz
        while not rospy.is_shutdown():
            self.command_pub.publish(msg)
            rate.sleep()


    def joint_states_callback(self, msg):
        self.joint_states = msg
        self.ready = True



if __name__ == '__main__':
    print('starting allegro_hand_initial')
    rospy.init_node('allegro_hand_initial', disable_signals=True)
    node = AllegroHandInitialNode()

    try:
        pass
        #rospy.spin()
    except:
        print('caught exception')
    print('exiting')

File: allegro_hand_teleop/src/allegro_hand_initial.cpp
#include "allegro_hand_teleop/allegro_hand_initial.h"

AllegroHandInitialNode::AllegroHandInitialNode(){
	// Initialize Variables
	ready = false;
	finished = false;

	// ROS Infrastructure
	joint_state_sub = nh.subscribe(JOINT_STATE_TOPIC, 1, &AllegroHandInitialNode::jointStateCallback, this);
	command_pub = nh.advertise<std_msgs::Float64MultiArray>(COMMAND_TOPIC, 5);

	// Wait to get ready
	ROS_INFO("Waiting for joint_states...");
	while(!ready)
		ros::spinOnce();
	ROS_INFO("Received first joint_states message");

	// Retrieve initial position
	command_msg.data = current_joint_state.position;

	std::vector<std::string> v;
	// Send initial goal
	while (ros::ok()){
		command_pub.publish(command_msg);
		ros::spinOnce();

		// ros::master::getNodes(v);
		// for (auto obj: v)
		// 	std::cout<<obj<<"\n";
		// std::cout<<std::endl;
		// if (std::find(v.begin(), v.end(), "/allegro_hand_hardware_interface") != v.end())
		// 	break;
	}
}

AllegroHandInitialNode::~AllegroHandInitialNode(){
	nh.shutdown();
}

void AllegroHandInitialNode::jointStateCallback(const sensor_msgs::JointState::ConstPtr& msg){
	current_joint_state = *msg;
	ready = true;
}

int main(int argc, char** argv)
{
	ros::init(argc, argv, "allegro_hand_initial");

	AllegroHandInitialNode node;
	//ros::spin();

	return 0;
}